plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java

```diff
@@ -35,121 +35,189 @@
+                if (storage == null) {
+                    s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
+                } else {
+                    s_logger.debug(String.format("Libvirt storage pool [%s] was found, but it is not running, removing it from HA list.", uuid));
                 }
+
+                removedPools.add(uuid);
+            }
+
+            s_logger.debug(String.format("Found NFS storage pool [%s] in libvirt, continuing.", uuid));
+
+        } catch (LibvirtException e) {
+            s_logger.debug(String.format("Failed to lookup libvirt storage pool [%s].", uuid), e);
+
+            if (e.toString().contains("pool not found")) {
+                s_logger.debug(String.format("Removing pool [%s] from HA monitor since it was deleted.", uuid));
+                removedPools.add(uuid);
+            }
+
+        }
+
+        String result = null;
+        for (int i = 1; i <= _heartBeatUpdateMaxTries; i++) {
+            Script cmd;
+            if (nfsStoragePool != null) {
+                cmd = createHeartBeatCommand(nfsStoragePool, hostPrivateIp, true);
+                result = cmd.execute();
+                s_logger.debug(String.format("The command [%s], to the pool [%s], had the result [%s].", cmd.toString(), uuid, result));
+            } else if (rbdStoragePool != null) {
+                cmd = createRbdHeartBeatCommand(rbdStoragePool, hostPrivateIp, true);
+                result = cmd.execute();
+                s_logger.debug(String.format("The command [%s], to the pool [%s], had the result [%s].", cmd.toString(), uuid, result));
             }
 
-            if (!removedPools.isEmpty()) {
-                for (String uuid : removedPools) {
-                    removeStoragePool(uuid);
+            if (result != null) {
+                s_logger.warn(String.format("Write heartbeat for pool [%s] failed: %s; try: %s of %s.", uuid, result, i, _heartBeatUpdateMaxTries));
+                try {
+                    Thread.sleep(_heartBeatUpdateRetrySleep);
+                } catch (InterruptedException e) {
+                    s_logger.debug("[IGNORED] Interrupted between heartbeat retries.", e);
                 }
+            } else {
+                break;
+            }
+
+        }
+
+        if (result != null && rebootHostAndAlertManagementOnHeartbeatTimeout) {
+            s_logger.warn(String.format("Write heartbeat for pool [%s] failed: %s; stopping cloudstack-agent.", uuid, result));
+            Script cmd;
+            if (nfsStoragePool != null) {
+                cmd = createHeartBeatCommand(nfsStoragePool, null, false);
+                result = cmd.execute();
+            } else if (rbdStoragePool != null) {
+                cmd = createRbdHeartBeatCommand(rbdStoragePool, null, false);
+                result = cmd.execute();
             }
         }
 
+        return removedPools;
```

Comment:
You are not doing anything with the returned set in line 125. So `return` is not needed